### PR TITLE
Feature/compare testrail and local tests structure

### PR DESCRIPTION
**Pull request by vladimir-s-quantory**
Added algorithm for getting or creating sectionId for .feature files in selected folder

### CODE REVIEW

**Review by SergeyRuzin — APPROVED** [review]

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/TestRailSynchronizer/TestRailSynchronizer.cs [hunk 1]
```diff
@@ -35,11 +39,12 @@ public void Sync(List<IFeatureFile> featureFiles)
                 foreach (var scenario in featureFile.Document.Feature.Children.OfType<Scenario>())
                 {
                     var tagId = scenario.Tags.FirstOrDefault(tag => Regex.Match(tag.Name, Config.TagIdPattern, RegexOptions.IgnoreCase).Success);
-                    //TODO: section selection logic
-                    ulong sectionId = 2197;
-                    //TODO: template selection logic
+                    //section selection logic
+                    ulong suiteId = 77;
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/TestRailSynchronizer/TestRailSynchronizer.cs:43]: Let's make it configurable, but taking a first suite by default?

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
+﻿using System;
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:1]: Clean up, please

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using GherkinSyncTool.Models;
+using GherkinSyncTool.Synchronizers.TestRailSynchronizer.TestRailManager;
+using GherkinSyncTool.Synchronizers.TestRailSynchronizer.TestRailManager.Model;
+using NLog;
+using TestRail.Types;
+
+namespace GherkinSyncTool.Synchronizers.SectionsSynchronizer
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:13]: Let's move to `GherkinSyncTool/Synchronizers/TestRailSynchronizer` namespace?

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        }
+        
+        /// <summary>
+        /// Builds a tree structure for TestRail sections
+        /// </summary>
+        /// <param name="projectId">TestRail project Id</param>
+        /// <param name="suiteId">TestRail suite Id</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException"></exception>
+        public IEnumerable<TestRailSection> GetSectionsTree(ulong projectId, ulong? suiteId)
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:32]: Looks like this is internal stuff and can be made private.

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+        /// <summary>
+        /// Gets or creates TestRail section Id for selected .feature file
+        /// </summary>
+        /// <param name="path">Path to .feature file</param>
+        /// <param name="suiteId">TestRail suite Id</param>
+        /// <param name="projectId">TestRail project Id</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public ulong GetOrCreateSectionId(string path, ulong suiteId, ulong projectId)
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:66]: I think it could be just `GetSectionId`

Comment (Brrovko) [reply GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:66]: `projectId ` argument can be removed. id can be taking from configs.

File: GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        {
+            var targetSections = GetSectionsTree(projectId, suiteId);
+            //Path includes name of the feature file and files catalog - hence SkipLast(1) and Skip(1)
+            var sourceSections = new Queue<string>(path.Split('\\').SkipLast(1).Skip(1));
+            var sectionId = GetOrCreateSectionIdRecursively(targetSections, sourceSections, suiteId, projectId);
+            if (sectionId is null)
+                throw new ArgumentNullException(nameof(sectionId), 
+                    "Section id cannot be null. You cannot put feature files in root folder");
```
Comment (Brrovko) [inline GherkinSyncTool/GherkinSyncTool/Synchronizers/SectionSynchronizer.cs:74]: Is it possible that `_testRailClientWrapper.CreateSection()` method returns null?

**Review by Brrovko — CHANGES_REQUESTED** [review]

**Review by SergeyRuzin — APPROVED** [review]

**Review by SergeyRuzin — APPROVED** [review]

**Pull request merged**